src/crdt.rs

```diff
@@ -0,0 +1,315 @@
+//! The `crdt` module defines a data structure that is shared by all the nodes in the network over
+//! a gossip control plane.  The goal is to share small bits of of-chain information and detect and
+//! repair partitions.
+//!
+//! This CRDT only supports a very limited set of types.  A map of PublicKey -> Versioned Struct.
+//! The last version is always picked durring an update.
+
+use bincode::{deserialize, serialize};
+use byteorder::{LittleEndian, ReadBytesExt};
+use hash::Hash;
+use result::Result;
+use ring::rand::{SecureRandom, SystemRandom};
+use signature::{PublicKey, Signature};
+use std::collections::HashMap;
+use std::io::Cursor;
+use std::net::{SocketAddr, UdpSocket};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, RwLock};
+use std::thread::{sleep, spawn, JoinHandle};
+use std::time::Duration;
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct Data {
+    key: PublicKey,
+    sig: Signature,
+    /// should always be increasing
+    version: u64,
+    /// address to connect to for gossip
+    gossip: SocketAddr,
+    /// address to connect to for replication
+    replicate: SocketAddr,
+    /// address to connect to when this node is leader
+    lead: SocketAddr,
+    /// current leader identity
+    who_leads: PublicKey,
+    /// last verified hash that was submitted to the leader
+    last_verified_hash: Hash,
+    /// last verified count, always increasing
+    last_verified_count: u64,
+}
+impl Data {
+    pub fn new(key: PublicKey, gossip: SocketAddr) -> Data {
+        let daddr = "0.0.0.0:0".parse().unwrap();
+        Data {
+            key,
+            sig: Signature::default(),
+            version: 0,
+            gossip,
+            replicate: daddr,
+            lead: daddr,
+            who_leads: PublicKey::default(),
+            last_verified_hash: Hash::default(),
+            last_verified_count: 0,
+        }
+    }
+    fn verify_sig(&self) -> bool {
+        //TODO implement this
+        true
+    }
+}
+
+/// `Crdt` structure keeps a table of `Data` structs
+/// # Properties
+/// * `table` - map of public key's to versioned and signed Data structs
+/// * `local` - map of public key's to self.table was updated
```

Comment:
Not sure I fully understand the function of the local map.  It's the value of the current node's index when the PublicKey data was examined?